pvlib/irradiance.py

```diff
@@ -833,7 +838,33 @@ def haydavies(surface_tilt, surface_azimuth, dhi, dni, dni_extra,
     sky_diffuse = dhi * (AI * Rb + term1 * term2)
     sky_diffuse = np.maximum(sky_diffuse, 0)
 
-    return sky_diffuse
+    # we've preserved the input type until now, so don't ruin it!
+    # if isinstance(sky_diffuse, pd.Series):
+    #     sky_diffuse[np.isnan(surface_tilt)] = 0
+    # else:
+    #     sky_diffuse = np.where(np.isnan(surface_tilt), 0, sky_diffuse)
+
+    if return_components:
+        diffuse_components = OrderedDict()
+        diffuse_components['sky_diffuse'] = sky_diffuse
+
+        # Calculate the individual components
+        diffuse_components['isotropic'] = np.maximum(dhi * term1 * term2, 0)
+        diffuse_components['circumsolar'] = np.maximum(dhi * (AI * Rb), 0)
```

Comment:
Rather than recalculating these quantities, I suggest storing each component in a variable (around line 837 or so),  calculating the overall `sky_diffuse` as the sum of those component variables, and then referencing those component variables here.  Calculating things only once is faster (probably not a big concern here) and lessens the risk of inconsistencies.